Provide backoff for retryable aws requests and the option to adjust the cf status check interval via an environment variable

philiiiiiipp opened the pull request:
## What did you implement

Closes #6980 

I did create that inside the github console, given the simple logic I do believe that its fine. But those might also be famous last words!

**_Is this ready for review?:_** YES
**_Is it a breaking change?:_** NO

medikoo commented on lib/plugins/aws/provider/awsProvider.js:
```diff
@@ -243,16 +243,21 @@ class AwsProvider {
               ) {
+                const nextTryNum = numTry + 1;
+                const exponentialBackOffTime = Math.round(
+                  5000 + nextTryNum * Math.random() * 10000
```
Maybe we can simplify, and set retry timeout to something between 4 and 7 seconds.

I wouldn't also add back off, it doesn't seem justified (we do not wait for long going request to finish).

medikoo replied:
@philiiiiiipp have you seen this comment maybe?

medikoo commented on lib/plugins/aws/provider/awsProvider.js:
```diff
@@ -243,16 +243,21 @@ class AwsProvider {
                     [
-                      `Recoverable error occurred (${e.message}), sleeping for 5 seconds.`,
-                      `Try ${numTry + 1} of ${MAX_TRIES}`,
+                      `Recoverable error occurred (${e.message}), sleeping for ${exponentialBackOffTime / 1000} seconds.`,
```
Maybe let's change it to `sleeping for few seconds`

philiiiiiipp replied:
We could do that, but I would prefer to have the exact number when looking at the logs. Especially now that we would move away from 5 seconds to a variable amount.

If you do not have a strong opinion about it I would leave it as is.

medikoo replied:
We will not have fixed seconds number.. it'll be e.g. 4.234s, so it's a bit noisy.

Why do you find this output valuable? Specifically to confirm that jitter functionality works?

philiiiiiipp commented:
I somehow cannot answer on your comments anymore. I believe that last unresolved one was about the backoff. 

In my opinion its a good pattern to have, but having a jitter in combination with the ability to increase the interval at the other place would suffice for my use case at least. I am not that knowledgeable in the codebase though, so if you prefer not having an increasing backoff, I am cool with that too.

medikoo commented:
> In my opinion its a good pattern to have

Can you elaborate why you see it useful in that case?

To me backoff serves well cases where we deal with a long wait time, and there's no point to check often. While here we deal with cases where there are short requests, and issue is that in some cases they're issued at very same time, so with fixed frequency they clash. In such case jitter seems the only solution that could indeed help,

I would like to avoid adding functionalities that do not serve real purpose (do not complicate things when it's not justified)

philiiiiiipp commented:
Right, we are doing requests against an endpoint, which might be congested, not only because of sls, but in general because there are many other clients trying to reach it. So if you want to allow the endpoint to recover it makes sense to slowly back off until your retries are used up.

https://docs.aws.amazon.com/general/latest/gr/api-retries.html

Like I said I am not married to this, I am more eager to get the jitter and the custom frequency option in, in order to stop using our forked version. So if you feel that this makes more harm than good I will remove it and we can always revisit it at a later stage.

medikoo commented:
> So if you feel that this makes more harm than good I will remove it

My feeling is that adding extra 5s every time, in this specific case, will rather bring experience downgrade than upgrade.

I would also narrow wait range to something as 4-7s.  The way it's configured now is 5-10s, so again that will make waits noticeably longer from what we have in master.

Technically I think it'll be good to focus in this PR purely on adding jitter, and maintain interval time on similar level as we have now.

The other reason I'm reluctant to make any bigger changes here, is that I don't remember having any other reports, that concerns that retry logic. So ideal approach seems to fix it for you by adding a jitter, but at same time maintain similar interval times, to keep experience not different for all other framework users.

philiiiiiipp commented:
> I would also narrow wait range to something as 4-7s. The way it's configured now is 5-10s, so again that will make waits noticeably longer from what we have in master.
> 
> Technically I think it'll be good to focus in this PR purely on adding jitter, and maintain interval time on similar level as we have now.
> 
> The other reason I'm reluctant to make any bigger changes here, is that I don't remember having any other reports, that concerns that retry logic. So ideal approach seems to fix it for you by adding a jitter, but at same time maintain similar interval times, to keep experience not different for all other framework users.

Right, that makes sense, I adjusted the interval to 4-7 seconds and removed the increasing backoff.
